solidity/SolidityParser.g4

```diff
@@ -0,0 +1,861 @@
+ * The default mutability 'non-payable' is assumed if no mutability is specified.
+ */
+stateMutability
+    : Pure
+    | View
+    | Payable
+    ;
+
+/**
+ * An override specifier used for functions, modifiers or state variables.
+ * In cases where there are ambiguous declarations in several base contracts being overridden,
+ * a complete list of base contracts has to be given.
+ */
+overrideSpecifier
+    : Override (LParen overrides += identifierPath (Comma overrides += identifierPath)* RParen)?
+    ;
+
+/**
+ * The definition of contract, library, interface or free functions.
+ * Depending on the context in which the function is defined, further restrictions may apply,
+ * e.g. functions in interfaces have to be unimplemented, i.e. may not contain a body block.
+ */
+functionDefinition
+    locals[
+	bool visibilitySet = false,
+	bool mutabilitySet = false,
+	bool virtualSet = false,
+	bool overrideSpecifierSet = false,
+]
+    : Function (identifier | Fallback | Receive) LParen (arguments = parameterList)? RParen (
+        {!$visibilitySet}? visibility {$visibilitySet = true;}
+        | {!$mutabilitySet}? stateMutability {$mutabilitySet = true;}
+        | modifierInvocation
+        | {!$virtualSet}? Virtual {$virtualSet = true;}
+        | {!$overrideSpecifierSet}? overrideSpecifier {$overrideSpecifierSet = true;}
+    )* (Returns LParen returnParameters = parameterList RParen)? (Semicolon | body = block)
+    ;
+
+/**
+ * The definition of a modifier.
+ * Note that within the body block of a modifier, the underscore cannot be used as identifier,
+ * but is used as placeholder statement for the body of a function to which the modifier is applied.
+ */
+modifierDefinition
+    locals[
+	bool virtualSet = false,
+	bool overrideSpecifierSet = false
+]
+    : Modifier name = identifier (LParen (arguments = parameterList)? RParen)? (
+        {!$virtualSet}? Virtual {$virtualSet = true;}
+        | {!$overrideSpecifierSet}? overrideSpecifier {$overrideSpecifierSet = true;}
+    )* (Semicolon | body = block)
+    ;
+
+/**
+ * Definition of the special fallback function.
+ */
+fallbackFunctionDefinition
+    locals[
+	bool visibilitySet = false,
+	bool mutabilitySet = false,
+	bool virtualSet = false,
+	bool overrideSpecifierSet = false,
+	bool hasParameters = false
+]
+    : kind = Fallback LParen (parameterList { $hasParameters = true; })? RParen (
+        {!$visibilitySet}? External {$visibilitySet = true;}
+        | {!$mutabilitySet}? stateMutability {$mutabilitySet = true;}
+        | modifierInvocation
+        | {!$virtualSet}? Virtual {$virtualSet = true;}
+        | {!$overrideSpecifierSet}? overrideSpecifier {$overrideSpecifierSet = true;}
+    )* (
+        {$hasParameters}? Returns LParen returnParameters = parameterList RParen
+        | {!$hasParameters}?
+    ) (Semicolon | body = block)
+    ;
+
+/**
+ * Definition of the special receive function.
+ */
+receiveFunctionDefinition
+    locals[
```

Comment:
All these locals and semantic predicates seem to force single usage of some syntax. This doesn't seem to be a large "positive" given that the checks are for a specific target, and we've been saying that semantic checking should be done post parse. Perhaps we can just remove these at this point? That said, one can easily remove all these via a single Trash command that edits the grammar prior to generating a parser.

Reply:
Also though about it, will remove a semantics related locals and predicates

Reply:
removed local predicates and rewrote parameterized rules to remove parameters